trio/_tests/test_threads.py

```diff
@@ -842,3 +844,166 @@ def __bool__(self):
 
     with pytest.raises(NotImplementedError):
         await to_thread_run_sync(int, cancellable=BadBool())
+
+
+async def test_from_thread_reuses_task():
+    task = _core.current_task()
+
+    async def async_current_task():
+        return _core.current_task()
+
+    assert task is await to_thread_run_sync(from_thread_run_sync, _core.current_task)
+    assert task is await to_thread_run_sync(from_thread_run, async_current_task)
+
+
+async def test_recursive_to_thread():
+    tid = None
+
+    def get_tid_then_reenter():
+        nonlocal tid
+        tid = threading.get_ident()
+        return from_thread_run(to_thread_run_sync, threading.get_ident)
+
+    assert tid != await to_thread_run_sync(get_tid_then_reenter)
+
+
+async def test_from_thread_host_cancelled():
+    queue = stdlib_queue.Queue()
+
+    def sync_check():
+        from_thread_run_sync(cancel_scope.cancel)
+        try:
+            from_thread_run_sync(bool)
+        except _core.Cancelled:
+            queue.put(True)
+        else:
+            queue.put(False)
+
+    with _core.CancelScope() as cancel_scope:
+        await to_thread_run_sync(sync_check)
+
+    assert not cancel_scope.cancelled_caught
+    assert not queue.get_nowait()
+
+    with _core.CancelScope() as cancel_scope:
+        await to_thread_run_sync(sync_check, cancellable=True)
+
+    assert cancel_scope.cancelled_caught
+    assert await to_thread_run_sync(partial(queue.get, timeout=1))
```

Comment:
Hm, this suggests that `from_thread.run_sync` *does* raise Cancelled if the thread was abandoned. Where is that coming from? It can't be from the cancel scope, because the cancel scope doesn't exist anymore (in the general case). If it's a general "this thread was abandoned" marker, then that's mildly backcompat-breaking and I think I disprefer it regardless - "sync functions have no checkpoints and therefore can't be the cause of a Cancelled" is a nice and easy rule to remember and works better if we don't have corner-case exceptions to it.

I'm not sure how _async_ `from_thread.run` from an abandoned thread should work. Options I see are:
* run it in a system task in a cancelled cancel scope
* run it in a system task that isn't cancelled, like we did before this PR
* raise Cancelled immediately

I would prefer one of the first two, since the third is unlike anything done elsewhere in Trio (Cancelled is raised at checkpoints, entry to an async function is not a checkpoint) and makes it hard to hand off resources safely (if you 'own' a resource and you pass it to a task that begins `[async] with resource:`, then that's normally safe but wouldn't be with the new semantics because Cancelled could be raised before the recipient enters the context manager).

Regardless of what we choose, we need to document it carefully, because it's a subtle consideration.

Reply:
I think I could on board for "run in cancelled system task" (maybe catching and dropping any exceptions?) but documenting the nuances of "if you abandon your thread, your async functions are going to run in a weird different task" seems really hard. 

What do you think of re-using `RunFinishedError` or making a `TaskFinishedError` to raise immediately? This would at least not trample on the semantics of `Cancelled`, but it wouldn't help with the resource handoff issue (on the other hand, users must already handle `RunFinishedError` in the present case).

Reply:
I implemented the semantics I think you expected, and there were enough degrees of freedom in the test suite that it didn't need much adjustment. That makes me think that we should think of a way to (a) document the surprising abandon-to-system-task semantics and (b) anchor them in the test suite. I can't yet think of which behaviors really should be represented that way though.

Reply:
Yeah, on further consideration I agree my previous proposal is fairly wonky and we can probably do better. I think my preference would actually be to have `cancellable=True` threads use the old semantics: don't pass cancellation, don't run `from_thread.run` reentrant calls in the `to_thread.run_sync` task, just use a system task (that is not cancelled) like we did before this PR. Rationale:
* Making stuff fail that worked before isn't great for backcompat. If someone is using the explicit "abandon thread when I'm cancelled" flag, then they might specifically want the abandoned thread to continue operating; our attempt to interrupt it as soon as we regain control might be counterproductive.
* Since the `cancellable=True` thread explicitly might outlive the original task, any attempt it makes to call back into Trio is probably for something unrelated to the original task's context, so it's strange for that attempt to reflect the cancellation status of the original task.
* I imagine a conceptual distinction between `cancellable=False` threads being an "extension of the current task", vs `cancellable=True` being "fire-and-forget + retrieve the result if you're still around when it becomes available". It just feels wrong to propagate cancellation in the latter case.

Reply:
I was going to suggest a new, overriding kwarg `abandon_on_cancel`. That tells you what happens and when, and my multiprocessing library could use `kill_on_cancel` instead. `cancellable` would then hang around indefinitely, deprecated but with unchanged semantics.

Reply:
I'm equally happy with `abandon_on_cancel` as a name.